### ISSUE TITLE
[Repair]: control allows setting a 'parallel' value greater than the allowed maximum

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla-manager#3527]
SM version: 3.2
Cluster topology: 3 nodes in each rack
RF: 3

*Description*
`sctool control` allowed to set `parallel` to 4 which is greater than the maximum allowed value on this setup (3).

And this value (4) shows in the `progress` output.
So, at least it's confusing. 

```
Run:            8981220a-3d53-11ee-bf62-42010a8e0019
Status:         RUNNING
Start time:     17 Aug 23 23:12:29 UTC
Duration:       3m46s
Progress:       88%
Intensity:      1
Parallel:       4
Datacenters:
  - us-east1

```

The amount of actual repair jobs doesn't seem to surpass the maximum so it's probably only about the display value and the API.

IMO it makes sense to always show the actually used value. In this case it means that we should have trimmed the provided value to 3 or fail the corresponding `control` command with an "illegal value" error.

**Comment by Michal-Leszczynski** [comment]
Right now SM should follow it's docs, so the `single job per single hos at any given time` requirement should make it impossible to ever cross the parallel limit. It's worth to remember that different keyspaces might have different parallel limits (due to different RFs). So correct values for parallel would be in [0, max_for_ks(#nodes / RF)].
Is this correct?

**Comment by vladzcloudius** [comment]
> Right now SM should follow it's docs, so the `single job per single hos at any given time` requirement should make it impossible to ever cross the parallel limit. 

Yup. That's correct. 
And I believe SM effectively doesn't break the contract from our docs (I don't have any evidence of the opposite).

This issue is more about the API.
If the actual state right now is that we allow the maximum 'parallel' value as `max_for_ks(max_parallel(KS))` then we can close this GH issue.

However if we don't have such a check then we probably want to add it and then close this issue.

> It's worth to remember that different keyspaces might have different parallel limits (due to different RFs). So correct values for parallel would be in [0, max_for_ks(#nodes / RF)]. Is this correct?

Correct but with some small correction.

If in the `max_for_ks(#nodes / RF)` `#nodes` stands for a total number of nodes in the cluster and `RF` - the total replication in the cluster then this formula is not going to give you a correct value for KS configurations like:

` { 'class' : 'NetworkTopologyStrategy', 'DC1': '3', 'DC2': '0' }`

and, say, 6 nodes in each DC (total 12 nodes in the cluster).

So, 12 / 3 == 4
while the correct max 'parallel' is 6 / 3 = 2

And it's not only for RF=0.

The more precise "max parallel" formula for each KS is:

```
max_parallel = 0
foreach DC:
  max_parallel = min(max_parallel, #nodes(DC) / RF(DC))
```

**Comment by Michal-Leszczynski** [comment]
> However if we don't have such a check then we probably want to add it and then close this issue.

There is no such check right now - parallel can be set to 9999999, but SM does not validate that because `single job per single hos at any given time` takes care of limiting the parallelism. But it's true that it would be nice to have this validation.

> Correct but with some small correction.

That's true! Although the `--ignore-down-hosts` and `--host` flags can also impact the maximal parallelism. So the formula of "max parallel" for keyspace should look like:
```
max_parallel = 99999999999999
foreach DC:
  max_parallel = min(max_parallel, #filtered_nodes(DC) / smallest_repaired_replica_set_size(DC))
```
For example, there is a keyspace with ` { 'class' : 'NetworkTopologyStrategy', 'DC1': '3'}` and there are 5 nodes (A,B,C,D,E) in this DC. Normally this would mean, that we can perform only 1 repair job in parallel, but if node A is down and we run repair task with `--ignore-down-hosts`, then we can repair two replica sets in parallel (e.g. [B,C] and [D,E]) even though RF = 3.

**Comment by vladzcloudius** [comment]
> > However if we don't have such a check then we probably want to add it and then close this issue.
> 
> There is no such check right now - parallel can be set to 9999999, but SM does not validate that because `single job per single hos at any given time` takes care of limiting the parallelism. But it's true that it would be nice to have this validation.
> 
> > Correct but with some small correction.
> 
> That's true! Although the `--ignore-down-hosts` and `--host` flags can also impact the maximal parallelism. So the formula of "max parallel" for keyspace should look like:
> 
> ```
> max_parallel = 99999999999999
> foreach DC:
>   max_parallel = min(max_parallel, #filtered_nodes(DC) / smallest_repaired_replica_set_size(DC))
> ```
> 
> For example, there is a keyspace with ` { 'class' : 'NetworkTopologyStrategy', 'DC1': '3'}` and there are 5 nodes (A,B,C,D,E) in this DC. Normally this would mean, that we can perform only 1 repair job in parallel, but if node A is down and we run repair task with `--ignore-down-hosts`, then we can repair two replica sets in parallel (e.g. [B,C] and [D,E]) even though RF = 3.

Right. 
And, yes, `max_parallel` was supposed to be initialized to a very-large-number and not to 0 like I wrote.
`num_nodes_in_cluster` would represent a good initial value.
